module/zfs/spa.c

```diff
@@ -4588,27 +4589,42 @@ spa_export_common(char *pool, int new_state, nvlist_t **oldconfig,
 	 * reacquire the namespace lock, and see if we can export.
 	 */
 	spa_open_ref(spa, FTAG);
+
+	/*
+	 * Save the txg that is being synced.
+	 */
+	if (spa->spa_sync_on) {
+		sync_txg = spa_syncing_txg(spa);
+	}
+
 	mutex_exit(&spa_namespace_lock);
 	spa_async_suspend(spa);
 	if (spa->spa_zvol_taskq) {
 		zvol_remove_minors(spa, spa_name(spa), B_TRUE);
 		taskq_wait(spa->spa_zvol_taskq);
 	}
-	mutex_enter(&spa_namespace_lock);
-	spa_close(spa, FTAG);
 
-	if (spa->spa_state == POOL_STATE_UNINITIALIZED)
-		goto export_spa;
 	/*
 	 * The pool will be in core if it's openable, in which case we can
 	 * modify its state.  Objsets may be open only because they're dirty,
 	 * so we have to force it to sync before checking spa_refcnt.
+	 *
+	 * If the pool is syncing wait for it to complete. This could take long
+	 * or may never complete if the IO is suspended (spa_suspended = 1).
+	 * Hence, it is best done without holding the spa_namespace_lock to
+	 * avoid blocking other operations which need spa_namespace_lock.
 	 */
-	if (spa->spa_sync_on) {
-		txg_wait_synced(spa->spa_dsl_pool, 0);
+	if (spa->spa_state != POOL_STATE_UNINITIALIZED && spa->spa_sync_on) {
```

Comment:
`spa->spa_sync_on` should imply `spa->spa_state != POOL_STATE_UNINITIALIZED`.  What do you think about converting the state check to an `ASSERT` in the conditional to make this explicit.

Reply:
I guess that makes sense. I guess I had added that check there protect against a parallel "zpool export" on this zpool. But, I don't think we need it as this thread has a reference on it.

Reply:
I see, yes that makes sense.